zebra-state/src/service/finalized_state/disk_db.rs

```diff
@@ -266,6 +277,31 @@ impl DiskDb {
         self.db.iterator_cf(cf_handle, rocksdb::IteratorMode::End)
     }
 
+    /// Returns a forward iterator over keys with `prefix` in `cf_name`,
+    /// starting from the first key with `prefix`.
+    ///
+    /// This iterator ends after returning all the keys with `prefix`.
+    ///
+    /// TODO: make the iterator return high-level types rather than bytes
+    ///       add an iterator wrapper struct that does disk reads in a blocking thread (#2188)
+    pub fn prefix_iterator<P>(
+        &self,
+        cf_handle: &rocksdb::ColumnFamily,
+        prefix: P,
+    ) -> impl IntoIterator<Item = (Box<[u8]>, Box<[u8]>)> + '_
+    where
+        P: IntoDisk,
+        <P as IntoDisk>::Bytes: 'static,
+    {
+        let prefix = prefix.as_bytes();
+
+        // Despite what the RocksDB wiki says,
+        // prefix iterators sometimes return extra trailing entries
+        self.db
+            .prefix_iterator_cf(cf_handle, prefix.as_ref())
+            .filter(move |(key_bytes, _value_bytes)| key_bytes.starts_with(prefix.as_ref()))
```

Comment:
Thanks for catching this!

Prefix iteration requires setting a prefix extractor. But it also seems to cause database hangs, or tokio hangs, regardless of the database thread mode. So I've given up on prefix iteration.

Instead, I used a standard iterator. It's almost as efficient, because we can start at the start of the transactions, and use a Rust filter to stop iteration. See commit feaf30aff.

Reply:
(And following commits, it turns out that any multi-key iteration takes a lock, which eventually causes a hang.)

Reply:
So this ended up not using iterators at all, except for very restricted single-key cases, wrapped in specific functions.